ddl/cluster.go

```diff
@@ -140,13 +143,91 @@ func checkFlashbackCluster(w *worker, d *ddlCtx, t *meta.Meta, job *model.Job, f
 	return nil
 }
 
+func addToSlice(schema string, tableName string, tableID int64, nonFlashbackIDs, allIDs []int64) ([]int64, []int64) {
+	if filter.IsSystemSchema(schema) && !strings.HasPrefix(tableName, "stats_") {
+		nonFlashbackIDs = append(nonFlashbackIDs, tableID)
+	}
+	allIDs = append(allIDs, tableID)
+	return nonFlashbackIDs, allIDs
+}
+
+// GetFlashbackKeyRanges make keyRanges efficiently for flashback cluster when many tables in cluster,
+// The time complexity is O(nlogn).
+func GetFlashbackKeyRanges(sess sessionctx.Context, startKey kv.Key) ([]kv.KeyRange, error) {
+	schemas := sess.GetDomainInfoSchema().(infoschema.InfoSchema).AllSchemas()
+
+	// The semantic of keyRanges(output).
+	var keyRanges []kv.KeyRange
+
+	// The allPhysicalIDs stores all physical tables IDs in TiDB cluster.
+	// The nonFlashbackPhysicalIDs holds all tables that don't need to do flashback.
+	var allPhysicalIDs, nonFlashbackPhysicalIDs []int64
+
+	for _, db := range schemas {
+		for _, table := range db.Tables {
+			if !table.IsBaseTable() || table.ID > meta.MaxGlobalID {
+				continue
+			}
+			nonFlashbackPhysicalIDs, allPhysicalIDs = addToSlice(db.Name.L, table.Name.L, table.ID, nonFlashbackPhysicalIDs, allPhysicalIDs)
+			if table.Partition != nil {
+				for _, partition := range table.Partition.Definitions {
+					nonFlashbackPhysicalIDs, allPhysicalIDs = addToSlice(db.Name.L, table.Name.L, partition.ID, nonFlashbackPhysicalIDs, allPhysicalIDs)
+				}
+			}
+		}
+	}
+
+	slices.Sort(allPhysicalIDs)
+	slices.Sort(nonFlashbackPhysicalIDs)
+
+	prevPos := -1
+	if len(nonFlashbackPhysicalIDs) > 0 {
+		prevPos, _ = slices.BinarySearch(allPhysicalIDs, nonFlashbackPhysicalIDs[0])
```

Comment:
Could we archive the same goal without binary search?

Let's take an example after the two slices are sorted:

allPhysicalIDs: [2, 3, 4, 9, 15, 20]
nonFlashbackPhysicalIDs: [1, 5, 12]

We could use 2 iterators that always iterated at the minimum elements of the 2 slices:

```
[2, 3, 4, 9, 15, 20]
 ^
 |
iterA

[1, 5, 12]
 ^
 |
iterN
```
If iterA < iterN, we add the ID of iterA to the output set and move iterA forward.
If iterA > iterN, we start a new ID ranges, skip the ID of iterN and move iterN forward.

Reply:
Totally possible. We can even maintain:

```
type flashbackID struct {
  id int
  excluded bool
}
```

Then insert all ID into one array and sort them on fly. Keyranges can be constructed simply by stop when you met `id.excluded == true`.